libgit/autogit_node_wrappers.go

```diff
@@ -33,16 +36,196 @@ import (
+	repoName string) *repoNode {
+	return &repoNode{
+		Node:     n,
+		am:       am,
+		h:        h,
+		repoName: repoName,
+	}
+}
+
+func (rn *repoNode) dstDir() string {
+	var typeStr string
+	switch rn.h.Type() {
+	case tlf.Public:
+		typeStr = public
+	case tlf.Private:
+		typeStr = private
+	case tlf.SingleTeam:
+		typeStr = team
+	}
+
+	return path.Join(
+		autogitRoot, typeStr, string(rn.h.GetCanonicalName()))
+}
+
+func (rn *repoNode) populate(ctx context.Context) bool {
+	ctx = context.WithValue(ctx, ctxSkipPopulateKey, 1)
+	children, err := rn.am.config.KBFSOps().GetDirChildren(ctx, rn)
+	if err != nil {
+		rn.am.log.CDebugf(ctx, "Error getting children: %+v", err)
+		return false
+	}
+
+	h, err := rn.am.config.KBFSOps().GetTLFHandle(ctx, rn)
+	if err != nil {
+		rn.am.log.CDebugf(ctx, "Error getting handle: %+v", err)
+		return false
+	}
+
+	// If the directory is empty, clone it.  Otherwise, pull it.
+	var doneCh <-chan struct{}
+	cloneNeeded := len(children) == 0
+	ctx = context.WithValue(ctx, ctxReadWriteKey, 1)
+	if cloneNeeded {
+		doneCh, err = rn.am.Clone(
+			ctx, rn.h, rn.repoName, "master", h, rn.dstDir())
+	} else {
+		doneCh, err = rn.am.Pull(
+			ctx, rn.h, rn.repoName, "master", h, rn.dstDir())
+	}
+	if err != nil {
+		rn.am.log.CDebugf(ctx, "Error starting population: %+v", err)
+		return false
+	}
+
+	select {
+	case <-doneCh:
+		return true
+	case <-ctx.Done():
+		rn.am.log.CDebugf(ctx, "Error waiting for population: %+v", ctx.Err())
+		// If we did a clone, ask for a refresh anyway, so they will
+		// see the CLONING file at least.
+		return cloneNeeded
+	}
+}
+
+func (rn *repoNode) shouldPopulate() (bool, <-chan struct{}) {
+	rn.lock.Lock()
+	defer rn.lock.Unlock()
+	if rn.populated {
+		return false, nil
+	}
+	if rn.populatingInProgress != nil {
+		return false, rn.populatingInProgress
+	}
+	rn.populatingInProgress = make(chan struct{})
+	return true, rn.populatingInProgress
+}
+
+func (rn *repoNode) finishPopulate(populated bool) {
+	rn.lock.Lock()
+	defer rn.lock.Unlock()
+	rn.populated = populated
+	close(rn.populatingInProgress)
+	rn.populatingInProgress = nil
+}
+
+// ShouldRetryOnDirRead implements the Node interface for
+// repoNode.
+func (rn *repoNode) ShouldRetryOnDirRead(ctx context.Context) (
+	shouldRetry bool) {
+	if ctx.Value(ctxSkipPopulateKey) != nil {
+		return false
+	}
+
+	// Don't let this operation take more than a fixed amount of time.
+	// We should just let the caller see the CLONING file if it takes
+	// too long.
+	ctx, cancel := context.WithTimeout(ctx, populateTimeout)
+	defer cancel()
+
+	for {
+		doPopulate, ch := rn.shouldPopulate()
+		if !doPopulate && ch == nil {
+			return shouldRetry
+		}
+		// If it wasn't populated on the first check, always force the
+		// caller to retry.
+		shouldRetry = true
+
+		if doPopulate {
+			rn.am.log.CDebugf(ctx, "Populating repo node on first access")
+			shouldRetry = rn.populate(ctx)
+			rn.finishPopulate(shouldRetry)
```

Comment:
Yes, you're right, but it's ok to set `populated` to `true` in that case.  We've only timed out _waiting_ for the clone to finish, the actual clone should keep going in the background.  There should be no need to call `Clone()` again unless there's a failure during the clone itself, which we currently don't try to detect.  We might have to worry about that at some point I guess -- i.e., set `populated` back to `false` if the `Clone()` ended up failing in the background -- but I didn't think it was needed for v1.  Let me know if you feel differently.

I'll update the comment in `populate` with some of these details.

Reply:
Sounds good; thanks. If a clone fails, I assume next push to the repo would still trigger a pull rather than being stuck there until process exists? (that is, with 1424 in as well)

Reply:
Yeah exactly.  Or the next read by a different client or after a client restart.